### PR TITLE
Add support for setting up `SMP` on `ARM64` through `ACPI`

### PR DESCRIPTION
**Pull request by mogasergiu**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): `arm64`
 - Platform(s): `kvm`
 - Application(s): N/A

### Description of changes

Add support for setting up `SMP` on `ARM64` through `ACPI`. 

This PR depends on #772 #848 #908 #909 #910 and #911, therefore it includes.
### NOTE
This is part of a larger set of Pull Requests. To make things easier and to ensure that things build as expected, it is recommended that testing is done based on #912, which includes everything. The splitting has been done to ease the review process.

To make testing easier, use the script from #910 to build the Unikraft images. You will need an `OVMF BIOS` image first.

### CODE REVIEW

File: plat/kvm/arm/setup.c [hunk 1]
```diff
@@ -189,7 +196,7 @@ void __no_pauth _ukplat_entry(struct ukplat_bootinfo *bi)
 	if (unlikely(rc))
 		UK_CRASH("Failed to initialize bootstrapping CPU: %d\n", rc);
 
-#ifdef CONFIG_HAVE_SMP
+#if defined(CONFIG_HAVE_SMP)
```
Comment (michpappas) [inline plat/kvm/arm/setup.c]: Is there a reason for this change?

File: plat/kvm/arm/setup.c [hunk 2]
```diff
@@ -48,28 +48,51 @@
 
 smccc_conduit_fn_t smccc_psci_call;
 
-static void _dtb_get_psci_method(void *fdt)
+/* TODO: Write all PSCI stuff as a proper driver... */
```
Comment (michpappas) [inline plat/kvm/arm/setup.c]: If I may, I find this `TODO` unnecessary given that we are all aware of this, and that `psci` will be moved into `drivers` shortly (re-arch).

File: plat/drivers/include/gic/gic-v2.h [hunk 3]
```diff
@@ -35,6 +35,9 @@
 
 #include <gic/gic.h>
 
+/* GICv2 GICC register map size page aligned up according to the spec. */
```
Comment (michpappas) [inline plat/drivers/include/gic/gic-v2.h]: According to what spec?

File: plat/drivers/include/gic/gic-v3.h [hunk 4]
```diff
@@ -67,6 +67,9 @@
  */
 #define GICC_CTLR_EL1_EOImode_drop	(1U << 1)
 
+/* Default size from the specification */
```
Comment (michpappas) [inline plat/drivers/include/gic/gic-v3.h]: Which spec? Also clarify in the commit message please.

File: plat/drivers/gic/gic-common.c [hunk 5]
```diff
@@ -34,6 +34,7 @@
 #include <gic/gic.h>
 #include <gic/gic-v2.h>
 #include <gic/gic-v3.h>
+#include <uk/plat/common/acpi.h>
```
Comment (michpappas) [inline plat/drivers/gic/gic-common.c:37]: Needs to be conditionally included or else build breaks when ACPI is not enabled.

File: plat/drivers/gic/gic-v2.c [hunk 6]
```diff
@@ -39,6 +39,8 @@
 #include <uk/asm.h>
 #include <uk/arch/limits.h>
 #include <uk/plat/lcpu.h>
+#include <uk/plat/common/acpi.h>
```
Comment (michpappas) [inline plat/drivers/gic/gic-v2.c:43]: Include conditionally?

File: plat/drivers/gic/gic-v2.c [hunk 7]
```diff
@@ -490,9 +492,8 @@ static int gicv2_initialize(void)
 	return 0;
 }
 
-static int gicv2_do_probe(const void *fdt)
+static inline void gicv2_set_ops(void)
```
Comment (michpappas) [inline plat/drivers/gic/gic-v2.c]: Personally I find `inline` when used with `static` to be meaningful only in headers, plus compilers is smart enough nowadays to inline code when needed, but I don't have a strong opinion.

File: plat/drivers/gic/gic-v2.c [hunk 8]
```diff
@@ -509,6 +510,71 @@ static int gicv2_do_probe(const void *fdt)
 
 	/* Set driver functions */
 	gicv2_drv.ops = drv_ops;
+}
+
+#if defined(CONFIG_UKPLAT_ACPI)
+/* Default page-aligned up size for GICv2 CPU interface according to spec */
```
Comment (michpappas) [inline plat/drivers/gic/gic-v2.c]: Spec which

File: plat/drivers/gic/gic-v2.c [hunk 9]
```diff
@@ -509,6 +510,71 @@ static int gicv2_do_probe(const void *fdt)
...
+#define GICV2_GICC_MEM_SZ	0x2000
```
Comment (michpappas) [inline plat/drivers/gic/gic-v2.c]: Perhaps that should be in `gic-v2.h` together with `GICD_V2_MEM_SZ`?

File: plat/drivers/gic/gic-v2.c [hunk 10]
```diff
@@ -509,6 +510,71 @@ static int gicv2_do_probe(const void *fdt)
...
+
+	rc = acpi_get_gicc(&gicv2_drv);
+	if (unlikely(rc < 0))
+		return rc;
+
+	rc = acpi_get_gicd(&gicv2_drv);
+	if (unlikely(rc < 0))
+		return rc;
+
+	if (unlikely(gicv2_drv.dist_mem_size != GICD_V2_MEM_SZ))
```
Comment (michpappas) [inline plat/drivers/gic/gic-v2.c:563]: Sorry maybe I missed something, but under what circumstances can this condition happen?

File: plat/drivers/gic/gic-v3.c [hunk 11]
```diff
@@ -44,6 +44,8 @@
 #include <uk/bitops.h>
 #include <uk/asm.h>
 #include <uk/plat/lcpu.h>
+#include <uk/plat/common/acpi.h>
```
Comment (michpappas) [inline plat/drivers/gic/gic-v3.c:48]: Include conditionally

File: plat/drivers/gic/gic-v3.c [hunk 12]
```diff
@@ -580,6 +581,71 @@ static int gicv3_do_probe(const void *fdt)
...
+		acpi_madt_gicr_t *gicr;
+		acpi_subsdt_hdr_t *h;
+	} m;
+	acpi_madt_t *madt;
+	__sz off, len;
+
+	madt = acpi_get_madt();
+	UK_ASSERT(madt);
+
+	/* We do not count the Redistributor regions, instead we rely on QEMU
```
Comment (michpappas) [inline plat/drivers/gic/gic-v3.c]: No assumptions that we run on QEMU please. I understand your point in the comment, but I would rephrase to something more generic.

File: plat/drivers/gic/gic-v3.c [hunk 13]
```diff
@@ -589,42 +655,42 @@ static int gicv3_do_probe(const void *fdt)
 	/* Get address and size of the GIC's register regions */
 	r = fdt_get_address(fdt, fdt_gic, 0, &gicv3_drv.dist_mem_addr,
 				&gicv3_drv.dist_mem_size);
-	if (r < 0) {
+	if (unlikely(r < 0)) {
 		uk_pr_err("Could not find GICv3 distributor region!\n");
 		return r;
 	}
 
+	/* TODO: Redistributors may be spread out in different regions.
```
Comment (michpappas) [inline plat/drivers/gic/gic-v3.c:665]: Okay to add this here (and good point 👍🏼) but IMO unnecessary to mention this addition in the commit message.

File: plat/common/arm/lcpu.c [hunk 14]
```diff
@@ -92,7 +93,59 @@ int lcpu_arch_init(struct lcpu *this_lcpu)
 	return ret;
 }
 
-int lcpu_arch_mp_init(void *arg)
+#ifdef CONFIG_UKPLAT_ACPI
+static int _lcpu_arch_mp_init(void *arg __unused)
```
Comment (michpappas) [inline plat/common/arm/lcpu.c]: As per our coding conventions, no leading underscores for `static` functions please. Prefer something else like `do_arch_mp_init()` or so.

**Review by michpappas — CHANGES_REQUESTED** [review]
@mogasergiu nice thanks, only minor comments on this one 🚀

File: plat/kvm/arm/setup.c (hunk 1 above)
Comment (mogasergiu) [reply plat/kvm/arm/setup.c]: An orphan change, I will delete it. If you were to look through the changelog, you would see that before there was a `#if defined(CONFIG_HAVE_SMP) && defined(CONFIG_UKPLAT_ACPI)`. But I removed it at some point and this change remained as a leftover from rebasing. Nice catch!

File: plat/drivers/gic/gic-common.c (hunk 5 above)
Comment (mogasergiu) [reply plat/drivers/gic/gic-common.c:37]: This happened with some x86 specific code as well, but I caught it in time. Ashamed that I did not catch it for the superior arch as well :stuck_out_tongue:.

File: plat/drivers/gic/gic-v2.c (hunk 7 above)
Comment (mogasergiu) [reply plat/drivers/gic/gic-v2.c]: I don't have one either, maybe I will revert it, maybe I won't :man_dancing:, depends on the mood. On a serious note though, I agree, it's in the same file and it is one less keyword.

File: plat/drivers/gic/gic-v2.c (hunk 10 above)
Comment (mogasergiu) [reply plat/drivers/gic/gic-v2.c:563]: This is because `GICC` entries exist for `GICv3` as well, so `GICv2`'s `acpi_get_gicc` may find `GICC`'s but they are not for the `GICv2`. Now that I take a look at it, since the `GICD` is common among both version, it should be checked for in `plat/drivers/gic/gic-common.c` and then decide upon which `GIC` version to actually probe. However, this cannot be done because `Devicetree` probing must also be accommodated for.... unless we assume `GICD` register map sizes between versions, which I would rather not do. The whole initial `_dtb_init_gic` looks dumb to me.

So, since the `GICD` is common for both, I would make the `gicd` pointer in `acpi_get_gicd` a `static` pointer so that we do not have to search for it a second time if both interrupt controllers end up being enumerated. Furthermore, I would make sure that `acpi_get_gicd` is called first in the `ACPI` probing. Let me know your thoughts.

Comment (mogasergiu) [reply plat/drivers/gic/gic-v2.c:563]: Actually, why can both versions exist in the first place? This is contrary to `ACPI`'s `GICD` structure:
> One, and only one, GIC distributor structure must be present in the MADT for an ARM based system

Why aren't `LIBGICV2` and `LIBGICV3` mutually exclusive?

File: plat/drivers/gic/gic-v3.c (hunk 13 above)
Comment (mogasergiu) [reply plat/drivers/gic/gic-v3.c:665]: But I do like being mean in commit messages :wine_glass::moyai:. On a serious note though, the reason why I thought it should be mentioned in the commit message is because otherwise it looks like an unrelated, orphan change. So I still think it is somewhat necessary, an additional commit message line does not hurt anybody IMHO. So, if you don't mind, I would like to keep it there :).

Comment (michpappas) [reply plat/drivers/gic/gic-v3.c:665]: Okay 👍🏼

File: plat/drivers/gic/gic-v2.c (hunk 10 above)
Comment (michpappas) [reply plat/drivers/gic/gic-v2.c:563]: > This is because GICC entries exist for GICv3 as well, so GICv2's acpi_get_gicc may find GICC's but they are not for the GICv2

Can you have a system with both a GICv2 and a GICv3? I haven't see one so far.

> Why aren't LIBGICV2 and LIBGICV3 mutually exclusive?

It's not because you can have a system with > 1 distributors. It's there because it allows a single build to be tested, released, and deployed for different machine types, that use GICv2 and GICv3.

Comment (mogasergiu) [reply plat/drivers/gic/gic-v2.c:563]: > Can you have a system with both a GICv2 and a GICv3? I haven't see one so far.

Not to my knowledge, that is why I asked why aren't the two configs mutually exclusive. What I meant is that if both are enabled, the GICv2 is probed first and it will look for GICC entries, which do not give you the GIC version, so the ACPI probe of GICv2 cannot know the version until the GICD is probed. That is why I said that I should have probably checked for the GICD first and also cache its result in a function locally declared `static` variable, so that when GICv3 is probed, it can fetch the result from when GICv2 was probed.

> It's not because you can have a system with > 1 distributors. It's there because it allows a single build to be tested, released, and deployed for different machine types, that use GICv2 and GICv3.

Ah, yes, it makes sense then. However, what drivers are (de)selected should not be reflected in the core codebase, this should have been transparently probed for in a generic, properly written, platform bus driver. But it's a discussion for later.

> So, since the GICD is common for both, I would make the gicd pointer in acpi_get_gicd a static pointer so that we do not have to search for it a second time if both interrupt controllers end up being enumerated. Furthermore, I would make sure that acpi_get_gicd is called first in the ACPI probing.

Unless you have any objections, I would proceed as I previously suggested.

File: plat/kvm/arm/setup.c [hunk 15]
```diff
@@ -48,28 +48,51 @@
...
+}
+#else
+static int get_psci_method(struct ukplat_bootinfo *bi)
 {
-	int fdtpsci, len;
 	const char *fdtmethod;
+	int fdtpsci, len;
+	void *fdt;
+
+	fdt = bi->dtb;
```
Comment (razvand) [inline plat/kvm/arm/setup.c]: Should be `fdt = (void *) bi->dtb`. Clang complains otherwise.

**Review by razvand — CHANGES_REQUESTED** [review]

**Comment by razvanvirtan** [comment]
@mogasergiu I looked on the code and tested with `nginx`, things look good to me :+1: 

One minor comment: in the https://github.com/unikraft/unikraft/pull/912/commits/980154623623641eecbba106d0791172ad8f8d04 commit message, you have one repetition: "fetch the `GIC Distributor`'s `GIC Distributor`'s physical base"

**Review by razvanvirtan — APPROVED** [review]
This looks good to me, thanks for the great work @mogasergiu!

Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

**Review by michpappas — APPROVED** [review]
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
